File: src/agent_memory/controller/forgetting.py
from __future__ import annotations

from dataclasses import dataclass
from math import exp, log

from agent_memory.models import MemoryItem, MemoryLayer


@dataclass(slots=True)
class ForgettingPolicy:
    short_term_beta: float = 1.2
    long_term_beta: float = 0.8
    promote_threshold: float = 0.7
    demote_threshold: float = 0.3

    def effective_strength(self, memory: MemoryItem, age_days: float) -> float:
        access_boost = 1 + log(1 + max(memory.access_count, 0))
        beta = self.long_term_beta if memory.layer is MemoryLayer.LONG_TERM else self.short_term_beta
        temporal_decay = exp(-memory.decay_rate * (age_days**beta))
        return memory.importance * memory.trust_score * access_boost * temporal_decay

    def next_layer(self, memory: MemoryItem, age_days: float) -> MemoryLayer:
        strength = self.effective_strength(memory, age_days=age_days)
        if strength >= self.promote_threshold:
            return MemoryLayer.LONG_TERM
        if strength <= self.demote_threshold:
            return MemoryLayer.SHORT_TERM
        return memory.layer

File: go-server/internal/controller/forgetting.go
package controller

import (
	"math"

	memoryv1 "github.com/bakebakebakebake/agent-memory/go-server/gen/memory/v1"
)

type ForgettingPolicy struct {
	ShortTermBeta    float64
	LongTermBeta     float64
	PromoteThreshold float64
	DemoteThreshold  float64
}

func NewForgettingPolicy() ForgettingPolicy {
	return ForgettingPolicy{
		ShortTermBeta:    1.2,
		LongTermBeta:     0.8,
		PromoteThreshold: 0.7,
		DemoteThreshold:  0.3,
	}
}

func (policy ForgettingPolicy) EffectiveStrength(memory *memoryv1.MemoryItem, ageDays float64) float64 {
	accessBoost := 1 + math.Log(1+math.Max(float64(memory.AccessCount), 0))
	beta := policy.ShortTermBeta
	if memory.Layer == "long_term" {
		beta = policy.LongTermBeta
	}
	temporalDecay := math.Exp(-memory.DecayRate * math.Pow(ageDays, beta))
	return memory.Importance * memory.TrustScore * accessBoost * temporalDecay
}

func (policy ForgettingPolicy) NextLayer(memory *memoryv1.MemoryItem, ageDays float64) string {
	strength := policy.EffectiveStrength(memory, ageDays)
	if strength >= policy.PromoteThreshold {
		return "long_term"
	}
	if strength <= policy.DemoteThreshold {
		return "short_term"
	}
	return memory.Layer
}
